rb_link/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+//! A Rust library for interacting with the Bluesim simulation program.
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, VecDeque};
+
+use std::io::{Read, Write};
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+use std::{fs, usize};
+
+#[cfg(test)]
+mod test;
+
+#[derive(Serialize, Deserialize)]
+pub enum GetPutMessage {
+    Get(u32),
+    Put(B2RMessage),
+}
+
+// pub struct ProbeInfo {
+//     id: u32,
+//     get_t_width: u32,
+//     put_t_width: u32,
+// }
+
+#[derive(Serialize, Deserialize)]
+struct R2BMessage {
+    id: u32,
+    message: Vec<u8>,
+}
+
+/// Message from Bluesim:
+/// - id: ID of the probe that sent the message
+/// - cycles: Clock cycles when the message was sent
+/// - message: Sent message, where message.len() == ceil(put_t_width/8). put_t_width is the width of put_t defined in your BSV code.
+#[derive(Serialize, Deserialize)]
+pub struct B2RMessage {
+    pub id: u32,
+    pub cycles: u32,
+    pub message: Vec<u8>,
+}
+
+/// A server for interacting with Bluesim.
+/// Cache bidirectional data and send data upon receiving requests.
+pub struct B2RServer {
+    // probe_infos: Arc<Mutex<HashMap<u32, ProbeInfo>>>,
+    b2r_cache: Arc<Mutex<HashMap<u32, VecDeque<B2RMessage>>>>,
+    r2b_cache: Arc<Mutex<HashMap<u32, VecDeque<R2BMessage>>>>,
+}
+
+impl Default for B2RServer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl B2RServer {
+    /// make a new server
+    pub fn new() -> B2RServer {
+        B2RServer {
+            // probe_infos: Arc::new(Mutex::new(HashMap::new())),
+            b2r_cache: Arc::new(Mutex::new(HashMap::new())),
+            r2b_cache: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+    // pub fn add_probe(&mut self, id: u32, get_t_width: u32, put_t_width: u32) {
+    //     let probe_info = ProbeInfo {
+    //         id,
+    //         get_t_width,
+    //         put_t_width,
+    //     };
+    //     self.probe_infos.lock().unwrap().insert(id, probe_info);
+    // }
+
+    /// Start a thread to run the server.
+    /// Create a UnixListener at "/tmp/b2rr2b".
+    /// Return the JoinHandle of that thread.
+    /// This function needs to be called before running your Bluesim program.
+    pub fn serve(&mut self) -> JoinHandle<()> {
+        // let probe_infos = self.probe_infos.clone();
+        let b2r_cache = self.b2r_cache.clone();
+        let r2b_cache = self.r2b_cache.clone();
+        thread::spawn(move || {
+            let _ = fs::remove_file("/tmp/b2rr2b");
+            let listener = UnixListener::bind("/tmp/b2rr2b").expect("Failed to bind Unix listener");
+            for stream in listener.incoming() {
+                let mut stream = match stream {
+                    Ok(stream) => stream,
+                    Err(_) => todo!(),
```

Comment:
error handling is important for a software. make sure you have handled all the error case.